ops/testing.py

```diff
@@ -0,0 +1,408 @@
+
+    The model that is created is from the viewpoint of the charm that you are testing.
+
+    :ivar charm: The instance of Charm that is backed by this testing harness.
+    :type charm: CharmBase
+    """
+
+    def __init__(self, charm_cls, meta=None):
+        """Used for testing your Charm or component implementations.
+
+        Example::
+
+            harness = Harness(MyCharm)
+            # Do initial setup here
+            relation_id = harness.add_relation('db', 'postgresql')
+            # Now instantiate the charm to see events as the model changes
+            harness.begin()
+            harness.add_relation_unit(relation_id, 'postgresql/0', remote_unit_data={'key': 'val'})
+            # Check that charm has properly handled the relation_joined event for postgresql/0
+            self.assertEqual(harness.charm. ...)
+
+        :param charm_cls: The Charm class that you'll be testing.
+        :param meta: (optional) A string or file-like object containing the contents of
+            metadata.yaml. If not supplied, we will look for a 'metadata.yaml' file in the
+            parent directory of the Charm, and if not found fall back to a trivial
+            'name: test-charm' metadata.
+        """
+        # TODO: jam 2020-03-05 We probably want to take config as a parameter as well, since
+        #       it would define the default values of config that the charm would see.
+        self._charm_cls = charm_cls
+        self._charm = None
+        self._meta = self._create_meta(meta)
+        self._unit_name = self._meta.name + '/0'
+        self._model = None
+        self._framework = None
+        self._hooks_enabled = True
+        self._relation_id_counter = 0
+        self._backend = _TestingModelBackend(self._unit_name)
+        self._model = model.Model(self._unit_name, self._meta, self._backend)
+        self._framework = framework.Framework(":memory:", "no-disk-path", self._meta, self._model)
+
+    @property
+    def charm(self):
+        return self._charm
+
+    @property
+    def model(self):
+        return self._model
+
+    @property
+    def framework(self):
+        return self._framework
+
+    def begin(self):
+        """Instantiate the Charm and start handling events.
+
+        Before calling begin(), changes to the model won't be
+        """
+        # The Framework adds attributes to class objects for events, etc. As such, we can't re-use
+        # the original class against multiple Frameworks. So create a locally defined class
+        # and register it.
+        # TODO: jam 2020-03-16 We are looking to changes this to Instance attributes instead of
+        #       Class attributes which should clean up this ugliness. The API can stay the same
+        class TestEvents(self._charm_cls.on.__class__):
+            pass
+
+        TestEvents.__name__ = self._charm_cls.on.__class__.__name__
+
+        class TestCharm(self._charm_cls):
+            on = TestEvents()
```

Comment:
you could also have called `type` to dynamically create a class with the right name (both for `TestEvents` and `TestCharm`), but i guess that's too arcane.
The above is super clear, if more verbose. I think you made the right call. :+1: